### PR TITLE
[IOTDB-3253] Upgrade dependency: Ratis from 2.2.0 to 2.3.0

### PR DESCRIPTION
**Pull request by SzyWilliam**
Recently Apache Ratis Community releases 2.3.0, which includes bugfix to known InstallSnapshot issues.
In this PR, I upgrade Ratis version to 2.3.0.

Notice:
  Currently, we name the snapshotDir with Term_Index so that we can tell which directory contains
  the latest snapshot files. Unfortunately, when leader install snapshot to a slow follower,
  current Ratis implementation will flatten the directory and place all the snapshots directly
  under statemachine dir. Under this scenario, we cannot restore Term_Index from directory name.
  We decided to add an empty metadata file containing only Term_Index into the snapshotDir. his
  metadata file will be installed along with application snapshot files, so that Term_Index
  information is kept during InstallSnapshot.

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java [hunk 1]
```diff
@@ -142,13 +144,22 @@ public long takeSnapshot() throws IOException {
     // require the application statemachine to take the latest snapshot
     String metadata = Utils.getMetadataFromTermIndex(lastApplied);
     File snapshotDir = snapshotStorage.getSnapshotDir(metadata);
+
+    // delete snapshotDir fully in case of last failed take snapshot
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:148]: typo

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java [hunk 2]
```diff
@@ -142,13 +144,22 @@ public long takeSnapshot() throws IOException {
...
     if (!snapshotDir.isDirectory()) {
       logger.error("Unable to create snapshotDir at {}", snapshotDir);
       return RaftLog.INVALID_LOG_INDEX;
     }
-    boolean success = applicationStateMachine.takeSnapshot(snapshotDir);
-    if (!success) {
+
+    boolean applicationTakeSnapshotSuccess = applicationStateMachine.takeSnapshot(snapshotDir);
+    boolean addTermIndexMetafileSuccess =
+        snapshotStorage.addTermIndexMetaFile(snapshotDir, metadata);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:159]: Do we need to clean up this if failed?

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java [hunk 3]
```diff
@@ -141,4 +137,75 @@ public File getStateMachineDir() {
...
+            "An unexpected condition triggered. please check implementation "
+                + this.getClass().getName());
+        FileUtils.deleteFully(snapshotDir);
+        return;
+      }
+
+      for (File file : snapshotFiles) {
+        boolean success = file.renameTo(new File(snapshotDir + File.separator + file.getName()));
+        if (!success) {
+          FileUtils.deleteFully(snapshotDir);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:203]: logger

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java (hunk 1 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:148]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java (hunk 2 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/ApplicationStateMachineProxy.java:159]: If takeSnapshot() failed, the whole `snapshotDir` will be deleted, including this TermIndexMetaFile.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java [hunk 4]
```diff
@@ -141,4 +137,75 @@ public File getStateMachineDir() {
...
+    return META_FILE_PREFIX + "\\d+_\\d+$";
+  }
+
+  /**
+   * After leader InstallSnapshot to a slow follower, Ratis will put all snapshot files directly
+   * under statemachineDir. We need to handle this special scenario and rearrange these files to
+   * appropriate sub-directory this function will move all snapshot files directly under /sm to
+   * /sm/term_index/
+   */
+  void rearrangeSMDirHierarchyIfNecessary() {
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:176]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java (hunk 3 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/SnapshotStorage.java:209]: done

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Pull request merged**